luacompiler/src/lib/irgen/instr.rs

```diff
@@ -0,0 +1,104 @@
+use irgen::opcodes::IROpcode;
+
+#[derive(PartialEq, Debug, Clone)]
+pub enum Arg {
+    Nil,
+    Int(i64),
+    Float(f64),
+    Str(String),
+    Reg(usize),
+    Func(usize),
+    Some(usize),
+}
+
+impl Arg {
+    pub fn get_reg(&self) -> usize {
+        if let Arg::Reg(reg) = self {
+            *reg
+        } else {
+            panic!("Arg was not a Reg; received {:?}", self)
+        }
+    }
+
+    pub fn get_some(&self) -> usize {
+        if let Arg::Some(some) = self {
+            *some
+        } else {
+            panic!("Arg was not a Some; received {:?}", self)
+        }
+    }
+
+    pub fn get_str(&self) -> String {
+        if let Arg::Str(s) = self {
+            s.clone()
+        } else {
+            panic!("Arg was not a Str; received {:?}", self)
+        }
+    }
+
+    pub fn get_func(&self) -> usize {
+        if let Arg::Func(f) = self {
+            *f
+        } else {
+            panic!("Arg was not a Func; received {:?}", self)
+        }
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub enum Instr {
```

Comment:
I'm not really sure what this represents: it seems to be the meta-representation of an instruction? Why do we want that vs. having actual instructions (add, sub, mul, call etc.)?

Reply:
An instructions has an opcode, and can take zero, one, two, three or n arguments. Before this, I had a struct which had an opcode and a vector of arguments.

Reply:
Yes, but I'm not sure why we're representing instructions at the meta-level. In other words, if I have an instruction `Add(a, b)`, at the meta-level, it's represented as something like `Instr("Add", [Var("a"), Var("b")])`.

Reply:
Then would something like:
```rust
enum Instr {
    Add {
        arg1: Arg,
        arg2: Arg,
        arg3: Arg,
    }
    Sub {
        arg1: Arg,
        arg2: Arg,
        arg3: Arg,
    }
    ...
}
```
be more suitable?

Reply:
I'm not saying which is more suitable, but I am asking you which is more suitable :)

Reply:
I am not sure which is more suitable. I would say my current solution involves less code duplication. If I were to use the above representation then I would have a need similar structs for all arithmetic operators. I could create a structure to represent an arithmetic operation, but I think  that will lead to unnecessary indirection.

Reply:
The honest answer is that I don't know and although my gut feeling vaguely suggests the opposite, it's not a very strong gut feeling. So I suggest you carry on down this route and see where it leads you!

Reply:
I just saw this in the PR channel and thought I'd chime in and say it might be worth looking at how rustc represents it's MIR (Mid-Level IR). MIR isn't in SSA form, but I don't think that's particularly relevant to this comment.

https://github.com/rust-lang/rust/blob/1349c84a4fa0fca9b866b2e859d28ee185ca0c1b/src/librustc/mir/mod.rs#L2190

They don't create instructions at the meta-level but they do create a single BinaryOp instruction and parameterize the actual binary operator. It might be overkill for your situation though

Reply:
I see, thank you! I will stick with my current design for now and see where it takes me. If I decide to switch to explicit instructions, I will probably represent binary operations in a similar way.